File: multithread-test/src/main/java/com/example/multithread_test/Counter.java
package com.example.multithread_test;

import java.util.concurrent.locks.ReentrantLock;

public class Counter {

    private int count = 0;
    private final ReentrantLock lock = new ReentrantLock();

    public void increment() {
        lock.lock();
        try {
            count++;
        } finally {
            lock.unlock();
        }
    }

    public int getCount() {
        lock.lock();
        try {
            return count;
        } finally {
            lock.unlock();
        }
    }
    public static void main(String[] args) throws InterruptedException {
        Counter counter = new Counter();

        // Create multiple threads to increment the counter
        Thread thread1 = new Thread(() -> {
            for (int i = 0; i < 10000; i++) {
                counter.increment();
            }
        });

        Thread thread2 = new Thread(() -> {
            for (int i = 0; i < 10000; i++) {
                counter.increment();
            }
        });

        thread1.start();
        thread2.start();

        // Wait for both threads to finish
        thread1.join();
        thread2.join();

        System.out.println("Final count: " + counter.getCount());
    }

}

File: job-scheduler-example/src/main/java/com/example/job_scheduler_example/Counter.kt
package com.example.job_scheduler_example

class Counter {
    // Shared counter variable (not thread-safe)
    var count = 0

    // Thread-safe increment using synchronized block
    fun incrementSynchronized() {
        synchronized(this) {
            // Lock on current object
            count++
        }
    }

    // Thread-safe increment using volatile (visibility guarantee)
    @Volatile
    var volatileCount = 0

    fun incrementVolatile() {
        volatileCount++ // Might not be immediately visible to other threads
    }
}

fun main() {
    val counter = Counter()

    // Spawn multiple threads to increment the counter
    val threads =
        List(10_000) {
            Thread {
                Thread.sleep(100)
                counter.incrementSynchronized()
            }
        }
    threads.forEach { it.start() }
    threads.forEach { it.join() }

    println("Synchronized counter: ${counter.count}") // May or may not be 10

    // Reset counter and use volatile
    counter.volatileCount = 0
    val volatileThreads =
        List(10_000) {
            Thread {
                Thread.sleep(100)
                counter.incrementVolatile()
            }
        }
    volatileThreads.forEach { it.start() }
    volatileThreads.forEach { it.join() }

    println("Volatile counter: ${counter.volatileCount}") // Might not be 10 (depends on visibility)
}
